clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp

```diff
@@ -0,0 +1,214 @@
+    StringRef Name) const noexcept {
+  if (llvm::any_of(ParsedCountingEnumPrefixes,
+                   [&Name](const StringRef &Prefix) -> bool {
+                     return Name.starts_with_insensitive(Prefix);
+                   }))
+    return true;
+  if (llvm::any_of(ParsedCountingEnumSuffixes,
+                   [&Name](const StringRef &Suffix) -> bool {
+                     return Name.ends_with_insensitive(Suffix);
+                   }))
+    return true;
+  return false;
+}
+
+size_t TaggedUnionMemberCountCheck::getNumberOfValidEnumValues(
+    const EnumDecl *Ed) noexcept {
+  bool FoundMax = false;
+  llvm::APSInt MaxTagValue;
+  llvm::SmallSet<llvm::APSInt, 32> EnumValues;
+
+  // Heuristic for counter enum constants.
+  //
+  //   enum tag_with_counter {
+  //     tag1,
+  //     tag2,
+  //     tag_count, <-- Searching for these enum constants
+  //   };
+  //
+  // The final tag count is decreased by 1 if and only if:
+  // 1. There is only one counting enum constant,
+  // 2. The counting enum constant is the last enum constant that is defined,
+  // 3. The value of the counting enum constant is the largest out of every enum
+  //    constant.
+  // The 'ce' prefix is a shorthand for 'counting enum'.
+  size_t CeCount = 0;
+  bool CeIsLast = false;
+  llvm::APSInt CeValue = llvm::APSInt::get(0);
+
+  for (const auto &Enumerator : Ed->enumerators()) {
+    const llvm::APSInt Val = Enumerator->getInitVal();
+    EnumValues.insert(Val);
+    if (FoundMax) {
+      if (greaterBySign(Val, MaxTagValue) ||
+          (signEquals(Val, MaxTagValue) && Val > MaxTagValue))
+        MaxTagValue = Val;
+    } else {
+      MaxTagValue = Val;
+      FoundMax = true;
+    }
+
+    if (EnableCountingEnumHeuristic) {
+      if (isCountingEnumLikeName(Enumerator->getName())) {
+        CeIsLast = true;
+        CeValue = Val;
+        CeCount += 1;
+        CountingEnumConstantDecl = Enumerator;
+      } else {
+        CeIsLast = false;
+      }
+    }
+  }
+
+  size_t ValidValuesCount = EnumValues.size();
+  if (CeCount == 1 && CeIsLast && CeValue == MaxTagValue) {
+    ValidValuesCount -= 1;
+  } else {
+    CountingEnumConstantDecl = nullptr;
+  }
+
+  return ValidValuesCount;
+}
+
+void TaggedUnionMemberCountCheck::check(
+    const MatchFinder::MatchResult &Result) {
+  const auto *Root = Result.Nodes.getNodeAs<RecordDecl>("root");
+  const auto *UnionField = Result.Nodes.getNodeAs<FieldDecl>("union");
+  const auto *TagField = Result.Nodes.getNodeAs<FieldDecl>("tags");
+
+  // The matcher can only narrow down the type to recordType()
+  if (!isUnion(UnionField))
+    return;
+
+  if (hasMultipleUnionsOrEnums(Root))
+    return;
+
+  const auto *UnionDef =
+      UnionField->getType().getCanonicalType().getTypePtr()->getAsRecordDecl();
+  const auto *EnumDef = llvm::dyn_cast_or_null<EnumDecl>(
+      TagField->getType().getCanonicalType().getTypePtr()->getAsTagDecl());
+
+  if (EnumDef) {
```

Comment:
```suggestion
  const auto *UnionDef =
      UnionField->getType().getCanonicalType().getTypePtr()->getAsRecordDecl();
  assert(UnionDef);
  const auto *EnumDef = llvm::dyn_cast_or_null<EnumDecl>(
      TagField->getType().getCanonicalType().getTypePtr()->getAsTagDecl());

  if (!EnumDef)
    return;
```

(And then pull the inner block of the current `if` one indentation out.)